core/trino-spi/src/main/java/io/trino/spi/block/BlockUtil.java

```diff
@@ -108,6 +108,41 @@ static int calculateBlockResetBytes(int currentBytes)
         return (int) newBytes;
     }
 
+    /**
+     * Returns an array containing elements in the specified range of the specified <code>isNull</code> array.
+     * If the input array is null, null is returned. If the range matches the entire array, the input array
+     * will be returned. Otherwise, a copy will be returned.
+     */
+    static boolean[] compactIsNull(@Nullable boolean[] isNull, int index, int length)
+    {
+        if (isNull == null) {
+            return null;
+        }
+        checkArrayRange(isNull, index, length);
+        if (index == 0 && length == isNull.length) {
+            return isNull;
+        }
+        int firstNull = 0;
+        for (; firstNull < length; firstNull++) {
+            if (isNull[firstNull + index]) {
+                break;
+            }
+        }
+        // If no nulls encountered, return null as the result
+        if (firstNull == length) {
+            return null;
+        }
+        boolean[] result = new boolean[length];
+        int offset = 0;
+        for (; offset < firstNull; offset++) {
+            result[offset] = false;
+        }
```

Comment:
The initial value of boolean array is false by default, we can skip it

Reply:
I thought this would be relevant for the JIT to elide a `memset` over the initial array during allocation, but after microbenchmarking it appears not to be and `System.arraycopy` performs better than the hand-rolled loops- so I've updated in the latest revision.